### DISCUSSION TITLE
User flows

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#149] [Anything Else]
"One of the main reasons for me to want Analytics is to see the users flow inside the Website itself, so that I can tailor it to the way the users are actually using it. The way I did it previously (when I had Google Analytics) was to have a Sankey Diagram which showed me how many users were flowing from one page to another. Is this something supported by Plausible or planned to be supported at some point?

Details:
For this to be useful for larger websites, I will describe how I would imagine it for a Blog (The web’s standard Hello World of websites :P )

The user would define custom page types. In this case, it would be:
HomePage: Has it’s own type
About: Has it’s own type
Posts: All posts share this type.

Then the Analytics allows me to select a page type (Say HomePage) and see how users flow from there (maybe 90% go to a Posts type page, and 10% to the About page).

Does this make sense?"

![ldud4xldqpwft8kfaovh](https://user-images.githubusercontent.com/34340819/85281041-9f0a6800-b489-11ea-960e-6fd25c00b748.png)

**UPDATE**: [User journeys / navigation flows](https://plausible.io/docs/user-journeys) are now available in Plausible!

**Comment by bradonomics** [comment]
This type of user flow data would be a prerequisite for moving people off Google Analytics who wanted to use predictive prefetching libraries like [Guess.js](https://guess-js.github.io/).

**Comment by Kiaka007** [comment]
Does anyone know if this has been addressed by Plausible? I agree that this feature is one of the main reasons I use analytics. It would be nice but doesn't have to be in that style. What I am looking for is to see where a visitor has entered the site (which is not the same as just visiting the page), exited the site, and where they went from what page.

**Reply by fdkaylo** [reply]
+1 For this. 

I actually dislike the design of this feature in Google Analytics. It should be nicer in Plausible.

I actually hesitate to buy Plausible due to this missing feature.

**Comment by dkoprowski** [comment]
👍 for this. I also created a discussion for this today: #695 but used different words "Navigation paths summary.

**Comment by supernrm** [comment]
+1, "must have"

**Comment by BrianNacho** [comment]
+1 Would be really helpful to have visualization for this

**Comment by Tobi-mmt** [comment]
+1 this is really essential to understand how the user navigates through the pages.

**Comment by bencresty** [comment]
+1
For me it doesn't have to be a visualization per se, but I'm 100% missing the User Explorer of GA, which does the same thing, but with a list of pages and the times the user was on these pages.

For me without this and a way to being able to see relations in the data (like relations between events, pages, etc., which is impossible now) it's getting to a point I will probably have to stop using Plausible after the trial period as it's just not enough to get an impression of usage. There is data yes, but it doesn't show a relation and so the data has too much gaps making some kind of useless for analizing IMO. Eventhough I really would like to go for Plausible and would be very happy to ditch GA and pay for a better party, it seems to be too early for my needs now. It's way too basic now. Hoping very much this will change soon.

**Comment by nick-phillips-dev** [comment]
This sounds like https://heap.io/

**Reply by supernrm** [reply]
No, it sounds like clicky.com (spy function)

**Reply by ukutaht** [reply]
My 2 cents:

I think we will build some sort of aggregated report for user flows at some point in the future. Sort of like the GA 'user flows' report although I'm not a fan of the UX there.

However, looking at individual user sessions and journeys is outside the scope of this project as a privacy-focused tool. Even the name `spy` in clicky hints at the privacy issues that come along with exposing individuals vs aggregated numbers in your analytics.

If you want to 'spy' on individual users or sessions I would first ask you to reconsider your analytics practices. I think recording and playing back visitor sessions should require explicit consent as it's stepping outside the visitors' reasonable expectation of online privacy. 

If you still think it's necessary to spy on user sessions, use tools that are built for that kind of analysis. Plausible is and will remain a tool that analyzes anonymous and aggregated data only.

**Reply by supernrm** [reply]
It's NOT spying WHAT user is doing on site (NOT HEAP, smartlook etc.), it's only WHAT PATH is going. Page A -> then B -> ten C and exit. I don't see here any privacy issues ;)

**Reply by bencresty** [reply]
> If you want to 'spy' on individual users or sessions I would first ask you to reconsider your analytics practices. I think recording and playing back visitor sessions should require explicit consent as it's stepping outside the visitors' reasonable expectation of online privacy.

@ukutaht At this time and age unfortunately we just can't expect privacy online, so visitors who still expect that are pretty naive IMO. That said, I'm all for privacy, do think it's very, and I mean VERY, important and I think it's great Plausible searches for ways to accomplish that. Although it seems to search for the right balance.

But lest's be realistic; what does Plausible contribute to online privacy and privacy as a wave if no company switches over to Plausible because these companies are missing useful analytics tools? So they stay with 'privacy unfriendly' tools like GA? Than it's missing the whole point of the change IMO and stays the same in the end. Companies just need to know how sites are used.

I agree with @supernrm in that being able to see what path users follow through a website is not in conflict with privacy. Or at least it shouldn't be.

Looks like to me an analytics tool that is putting privacy first is contradicting itself in the first place, so without a middle road what's the use of an analytics tool if we cannot get any insights from it about how our products are used?

When giving functional insights it isn't required to show age, gender, ip, address, profession, etc. like Google does. So leaving that behind is a huge step forward already I think.

I do think that Plausibile could be a little more realistic on what an analytics tool is actually used for by professionals; understanding what to improve about an online product, like a site or an app. Only seeing what pages are visited most just isn't telling the full story.

To me it's sounds a little naive, childish and easy to call this 'spying'. It's just information we need to understand what works and what doesn't, nothing more than that. I also think that if Plausible actually doesn't want to provide connected information so we can functionally analyze; why make an analytics tool in the first place? It seems to be in conflict with itself to me at this moment. I think the real question here is what the best balance is between privacy and analytics. To me it's at a point now that it makes really getting a clue of site-/app-usage guess work.

My 2 cts.

**Reply by metmarkosaric** [reply]
Thanks for your feedback! Privacy of visitors is something we don't want to compromise at any time now nor in the future. Professionals can get enough useful data from aggregate numbers and there is no need for any detail on individual level to make important decisions. If you need to get details on the personal level, it's best to ask the visitor for consent (and it's legally required in the EU at least) and if they give the consent use the tools that give you that level of individual detail. There are so many tools that can do that for you already and there's no need for another one. The reason we're working on Plausible is to give people a privacy-first and privacy regulation compliant way to get useful and actionable data.

**Reply by bencresty** [reply]
@metmarkosaric I wonder why you guys actively ask for our opinions if you're this stubborn and not discuss about actual real life usage requests coming in.

I wish you the best.

**Reply by metmarkosaric** [reply]
We're happy to discuss any requests and we even respond to the majority of messages as you can see yourself. Try to get a response from Google on the requests you send to them.

Plausible is a privacy first project so we don't plan to action any feedback nor request that compromises that. Nothing to do with being stubborn and everything to do with the ideals and the goals of the project. 

There are so many analytics tools that already collect all the data that's possible to collect that so there's certainly no need to build another one.

**Reply by bencresty** [reply]
> We're happy to discuss any requests and we even respond to the majority of messages as you can see yourself. Try to get a response from Google on the requests you send to them.
> 
> Plausible is a privacy first project so we don't plan to action any feedback nor request that compromises that. Nothing to do with being stubborn and everything to do with the ideals and the goals of the project.
> 
> There are so many analytics tools that already collect all the data that's possible to collect that so there's certainly no need to build another one.

Who are you talking to? You opened the request yourself and it has 10 upvotes and 12 likes at the moment. 
If you don't want it yourself and made up your mind before even embracing and discussing our responses, than don't open a post, don't ask for our opinions, close the discussion forum and just create the project how you like. That's also a choice.

I was so happy to have found a privacy consious party a few weeks ago and it looked like there was finally a way to leave GA and other products out there not respecting privacy. I thought and still think you have a chance to change the online privacy issues. But now I'm all dissapointed by all this weird conflicting communication, stubborness and unrealistic ways that making a real change is just unrealistic like this and we have to pay money for too little information that's not helping. Eventhough I was really looking forward to pay for it if we still have analytics and a great balance with privacy.
 
You guys have such a great opportunity to really make a change IMO, but like this I don't see it ever happen in the real world. It's just too idealistic, too soon. To me it's now either not having analytics, which is not an option, or back to the crazy privacy scandal parties, which is a great dissapointment as I don't want it, neither for myself, neither for my customers. And I'm pretty sure I'm not the only one thinking like this.

When you're in a shop there are camera's everywhere and that's ok, because everybody understands you're being watched and the reasons behind it. It only goes too far IMO if the camera's have facial tracking, putting you in a database with your name and address and your customer ID etc. But just the camera's are purely functional. This is to me exactly the same as with analytics.

But this is the last thing I wrote about anything Plausible.

**Reply by metmarkosaric** [reply]
As mentioned earlier in this thread, we plan to build some sort of aggregated report for user flows at some point in the future. This is what this thread is about. What we said no to is the idea to build a detailed user flow (like the "Spy" feature that's been mentioned) which allows you to analyze the behavior of individual visitors. That we don't want/plan to do. Thank you!

**Reply by ukutaht** [reply]
@bencresty The vision of this project is and has always been to be aggregate-only with no reporting on individual users or sessions.

You're free to disagree with our direction and you're free to fork the code and build the features you want. But you're not going to convince us to abandon a principle on which the whole project has been built so far.

>  Who are you talking to? You opened the request yourself and it has 10 upvotes and 12 likes at the moment.
If you don't want it yourself and made up your mind before even embracing and discussing our responses, than don't open a post, don't ask for our opinions, close the discussion forum and just create the project how you like. That's also a choice.

The discussion thread is about building a 'user flows' report. The industry standard (GA) does it in an aggregated way that doesn't show individual sessions. That's what the discussion is about. You changed the topic to showing individual sessions, which goes against our principles as I expressed.

Damn right I'm stubborn when it comes to principles, why else would I even be working on this product and giving away the code for free on Github?

**Reply by Tobi-mmt** [reply]
I don't understand the use case of seeing an individual flow. 
For me as an analytics user I want to know wehere the people overall enters my page, which path they go and where they exit. The most important part is where they exit, so I want to know e.g. if a user exits on page X , 30% of the users was on page Y before that. This kind of informations are important to me. I don't need a single journey of a single user.
So @metmarkosaric and @ukutaht from my point of view the aggregated numbers give analytics users enought informations to understand the flows of the users and to understand e.g. why they leave the page or how they came to a specific page.

And I would also agree connecting a specific user journey with the (hashed) ip is some sort of spying and needs a consent. 
_(@bencresty This is called pseudonymized data. But with enough effort you can reconstruct the IP. )_

But without this connection – only "count" the different journeys – there should be no privacy problem.
_(This is called anonymized data where you can not reconstruct the IP even if you have the whole power of the universe)_

**Reply by bencresty** [reply]
(also @metmarkosaric )
@ukutaht You're right and sorry for being so stubborn myself by my frustration.

In the end I believe we are actually thinking alike and we had a miscommunication here (I blame myself for that as I brought it up the wrong way). I didn't really meant to follow a single user, but meant to follow the way users follow between pages.
I also understand it's a privacy concern when connecting all data and I want to have a world with better privacy too, so I didn't mean to connect all data. Again, I put it the wrong way and it didn't came out as how I do think about it. Instead of connecting all data, I meant that IMO it is missing some relationships that makes data much more relevant.

On the other hand, you are completely right in your privacy concerns. And that's the whole reason I am looking at Plausible to begin with. So I was also in conflict with myself and have to get used to the idea that it's not possible to have both.

> @bencresty The vision of this project is and has always been to be aggregate-only with no reporting on individual users or sessions.

And that's the vision that drawn me to Plausible to start with.

> You're free to disagree with our direction and you're free to fork the code and build the features you want. But you're not going to convince us to abandon a principle on which the whole project has been built so far.

It's actually also great to see you are so convinced and invested in this. That's also really assuring and convincing that you're not going to compromise things in the future.

> The discussion thread is about building a 'user flows' report. The industry standard (GA) does it in an aggregated way that doesn't show individual sessions. That's what the discussion is about. You changed the topic to showing individual sessions, which goes against our principles as I expressed.

There was clearly a miscommunication, and again, I blame myself for that. What I meant to say in my first post was that for me this user flow doesn't have to be a fancy graph per se, but a textual list, like the GA user explorer, would be fine too. But than the discussion went to tracking individual users (which I followed and reacted to) and that wasn't what I meant with my post. If the users flow shows what flow users are following from a page to a page to a page that's exactly what I was after.

The only reason I don't used the user flow with GA is that, like you, I find that graphical interface not the best.

But to be clear: also when it's not possible to really see the full flow from page to page of a single user, which is something that takes some time to get used to for me, I'd rather use Plausible and have better privacy for visitors and be a little part of a change online, than staying with the big spying companies where everything is only about making money no matter what.

> Damn right I'm stubborn when it comes to principles, why else would I even be working on this product and giving away the code for free on Github?

You're right.

Sorry for all the inconvinience and the childish primitive way I handled this in my posts.

@Tobi-mmt 
> I don't understand the use case of seeing an individual flow.
> For me as an analytics user I want to know wehere the people overall enters my page, which path they go and where they exit. The most important part is where they exit, so I want to know e.g. if a user exits on page X , 30% of the users was on page Y before that. This kind of informations are important to me. I don't need a single journey of a single user.
> And I would also agree connecting a specific user journey with the (hashed) ip is some sort of spying and needs a consent.
> _(@bencresty This is called pseudonymized data. But with enough effort you can reconstruct the IP. )_
> 
> But without this connection – only "count" the different journeys – there should be no privacy problem.
> _(This is called anonymized data where you can not reconstruct the IP even if you have the whole power of the universe)_

I agree with you on this completely.

**Reply by ukutaht** [reply]
@bencresty great! Sorry if I misunderstood you or contributed to the miscommunication in some way :)

> The only reason I don't used the user flow with GA is that, like you, I find that graphical interface not the best.

right? It's so busy and overwhelming to me. Some ideas how to do better:

1. Just list the most common journeys through the whole website? Like

`/page1` -> `/page2` -> `/page3`   -- 2k visitors
`/page3` -> `/page1` -> `/page2`   -- 1k visitors

Maybe a of journeys would be easier to read than the graph in GA. Not sure.

2. When you click on a page to filter the dashboard down to a single page, we could show a list of `next page` and `previous page`. This should give a good idea how people come and go on certain pages in your app.

Just throwing around some ideas. I'm very interested in ideas on how to make a better 'user flows' report than what we can get from GA. Absent any better ideas we could also try to replicate GA in this regard.

**Comment by Semmu** [comment]
+1

This is a must-have feature if one wants to track user journeys, performance of new features, A/B testing, etc. I would really love to see this supported!

**Reply by Semmu** [reply]
After reading through the discussions here, I would like to share my ideas about a possible implementation.

First of all, I would not generate user journeys and reports automatically, because it would be too noisy, unnecessary for most cases, and also would probably impact the performance of the service. The need for the generation of these reports should remain explicit.

Also for the initial version I would not strive for a feature-rich implementation, rather for the smallest MVP which already brings value. So that shiny graph above is definitely out-of-scope.

What I think could work would be based on the existing goals defined. There we should be able to define special "pairs" of goals which we're interested in, the first one being an initial condition (must have), and the second being the related goal we want to measure the performance of.

Example: user hits the Pricing page, an event is fired, and then we want to measure what percentage of the visitors then went on to visit the Sign Up page, meaning we also fired that event in their journey.

The strictness of the system is up for debate, i.e. can they visit intermediate pages between these two or should they be strictly consecutive page loads, but I think both would be informative in their own way.

With this feature we would be easily able to measure the performance of custom A/B tests, new features, etc., which would satisfy the needs of most users.

**Reply by metmarkosaric** [reply]
thanks @Semmu! this is something along the same lines as we're thinking. there's also the thread for funnels #171

**Reply by Semmu** [reply]
Ohh I did not see that discussion, thanks for the heads up!

**Comment by GavinWalklate** [comment]
could it start of simple - like now and next - eg in dashboard you filter a single  page - and perhaps similar to the entrance and exits you can see the lists of pages that  users came from, and pages the user goes to (not a flow visualization of the entire user journey) but even this could provide some valuable insight - eg what page was a customer on before they came to our contact us page. From this maybe a visualization chart could also evolve. (to be clear this does not need to be a particular user journey)

**Comment by metmarkosaric** [comment]
**A related update**: We've now opened up a preview version of funnels to everyone. You can now follow the visitor journey from a landing page to a conversion with funnel analysis. Details: https://plausible.io/docs/funnel-analysis

**Comment by masteradhoc** [comment]
Adding my vote here also to get this into the existing subscriptions level of Plausible

**Comment by MickL** [comment]
This is a MUST for us. We are using etracker and we are not happy with its oldish approach (doesnt work well with Nuxt) BUT the features are way advanced atm. I can click on each user and see the full user journey:

1. Where he came from (Google Ads campagne name)
2. Where he navigated
3. How much time he spent on each page
4. Custom events he triggered

We really want to switch over to Plausible but this is a must have for us. I dont see Funnels to be useful for us at all.

**Comment by metmarkosaric** [comment]
**UPDATE**: [User journeys / navigation flows](https://plausible.io/docs/user-journeys) are now available in Plausible!
